output/elasticsearch/client.go

```diff
@@ -0,0 +1,476 @@
+// successfully added to the list of bulk items and the list of bulk items.
+func (client *Client) bulkEncodePublishRequest(version version.V, data []*messages.Event) ([]*messages.Event, []interface{}) {
+	okEvents := data[:0]
+	bulkItems := []interface{}{}
+	for _, event := range data {
+		meta, err := client.createEventBulkMeta(version, event)
+		if err != nil {
+			client.log.Errorf("Failed to encode event meta data: %+v", err)
+			continue
+		}
+		bulkItems = append(bulkItems, meta, beatsEventForProto(event))
+		okEvents = append(okEvents, event)
+	}
+	return okEvents, bulkItems
+}
+
+func GetOpType(e *messages.Event) events.OpType {
+	opStr, err := getMetaStringValue(e, events.FieldMetaOpType)
+	if err != nil {
+		return events.OpTypeDefault
+	}
+
+	switch opStr {
+	case "create":
+		return events.OpTypeCreate
+	case "index":
+		return events.OpTypeIndex
+	case "delete":
+		return events.OpTypeDelete
+	}
+
+	return events.OpTypeDefault
+}
+
+func (client *Client) createEventBulkMeta(version version.V, event *messages.Event) (interface{}, error) {
+	eventType := ""
+	/*if version.Major < 7 {
+		eventType = defaultEventType
+	}*/
+
+	pipeline := ""
+	/*pipeline, err := client.getPipeline(event)
+	if err != nil {
+		err := fmt.Errorf("failed to select pipeline: %v", err)
+		return nil, err
+	}*/
+
+	index := "elastic-agent-shipper"
+	/*index, err := client.index.Select(event)
+	if err != nil {
+		err := fmt.Errorf("failed to select event index: %v", err)
+		return nil, err
+	}*/
+
+	id, _ := getMetaStringValue(event, events.FieldMetaID)
+	opType := GetOpType(event)
+
+	meta := eslegclient.BulkMeta{
+		Index:    index,
+		DocType:  eventType,
+		Pipeline: pipeline,
+		ID:       id,
+	}
+
+	if opType == events.OpTypeIndex {
+		return eslegclient.BulkIndexAction{Index: meta}, nil
+	}
+	return eslegclient.BulkCreateAction{Create: meta}, nil
+}
+
+/*func (client *Client) getPipeline(event *messages.Event) (string, error) {
+	if event.GetMetadata() != nil {
+		pipeline, err := getMetaStringValue(event, events.FieldMetaPipeline)
+		if err == mapstr.ErrKeyNotFound {
+			return "", nil
+		}
+		if err != nil {
+			return "", errors.New("pipeline metadata is no string")
+		}
+
+		return strings.ToLower(pipeline), nil
+	}
+
+	if client.pipeline != nil {
+		return client.pipeline.Select(event)
+	}
+	return "", nil
+}*/
+
+// bulkCollectPublishFails checks per item errors returning all events
+// to be tried again due to error code returned for that items. If indexing an
+// event failed due to some error in the event itself (e.g. does not respect mapping),
+// the event will be dropped.
+func (client *Client) bulkCollectPublishFails(result eslegclient.BulkResult, data []*messages.Event) ([]*messages.Event, bulkResultStats) {
+	reader := newJSONReader(result)
+	if err := bulkReadToItems(reader); err != nil {
+		client.log.Errorf("failed to parse bulk response: %v", err.Error())
+		return nil, bulkResultStats{}
+	}
+
+	count := len(data)
+	failed := data[:0]
```

Comment:
I added the other one but I think this one makes sense as-is, there's no need to preallocate anything especially a full-length buffer since this array will usually (knock on wood 😜) be empty

Reply:
> I added the other one but I think this one makes sense as-is, there's no need to preallocate anything especially a full-length buffer since this array will usually (knock on wood 😜) be empty

The slice is pre-allocating, that was part of why I proposed the change, the make is a little more declarative.

See https://go.dev/play/p/JXwF_bFUYgE

And of course I screwed up in the suggestion, it should be `make([]*messages.Event, 0, len(data))`  I forgot the 0 for the length.  If we really expect failed to be empty we could do a make with len & cap set to 0.

Reply:
Sorry, could you elaborate on how an allocation is happening? As I understand it, `failed := data[:0]` creates a slice but doesn't allocate anything new, it just references the existing array `data` -- it should cause no heap allocations unless `append` is called, which creates a new underlying array if needed. I've seen this pattern used as a short way to create an auxiliary list that will often be empty, specifically because it avoids the allocation.